File: tools/performance/boundary.py
#!/usr/bin/env python3
"""Measure the installed Python package boundary with a portable CloudEvent round trip."""

from __future__ import annotations

import argparse
import json
import math
import time
from pathlib import Path

import forgelib


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--iterations", type=int, default=1000)
    parser.add_argument("--output")
    args = parser.parse_args()
    if args.iterations < 1:
        parser.error("iterations must be positive")
    event = {"id": "benchmark", "source": "urn:forge:performance", "type": "forge.benchmark", "data": b"boundary"}
    samples = []
    for _ in range(args.iterations):
        started = time.perf_counter()
        forgelib.decode_cloud_event(forgelib.encode_cloud_event(event))
        samples.append((time.perf_counter() - started) * 1000)
    samples.sort()
    rank = max(0, math.ceil(len(samples) * 0.95) - 1)
    report = {
        "schema_version": 1,
        "kind": "language_boundary",
        "language": "python",
        "iterations": args.iterations,
        "metrics": [{"name": "cloudevent_roundtrip_p95_ms", "value": samples[rank], "unit": "ms"}],
    }
    encoded = json.dumps(report, indent=2) + "\n"
    if args.output:
        Path(args.output).write_text(encoded)
    else:
        print(encoded, end="")


if __name__ == "__main__":
    main()

File: tools/performance/boundary.mjs
import { performance } from 'node:perf_hooks';
import { writeFileSync } from 'node:fs';
import { encodeCloudEvent, decodeCloudEvent } from '../../bindings/node/client.js';

const args = process.argv.slice(2);
const valueAfter = (name, fallback) => {
  const index = args.indexOf(name);
  return index === -1 ? fallback : args[index + 1];
};
const iterations = Number(valueAfter('--iterations', '1000'));
const output = valueAfter('--output', null);
if (!Number.isSafeInteger(iterations) || iterations < 1) throw new TypeError('iterations must be a positive integer');
const event = { id: 'benchmark', source: 'urn:forge:performance', type: 'forge.benchmark', data: Buffer.from('boundary') };
const samples = [];
for (let iteration = 0; iteration < iterations; iteration += 1) {
  const started = performance.now();
  decodeCloudEvent(encodeCloudEvent(event));
  samples.push(performance.now() - started);
}
samples.sort((left, right) => left - right);
const rank = Math.max(0, Math.ceil(samples.length * 0.95) - 1);
const report = {
  schema_version: 1,
  kind: 'language_boundary',
  language: typeof Bun === 'undefined' ? 'node' : 'bun',
  iterations,
  metrics: [{ name: 'cloudevent_roundtrip_p95_ms', value: samples[rank], unit: 'ms' }],
};
const json = `${JSON.stringify(report, null, 2)}\n`;
if (output) writeFileSync(output, json);
else process.stdout.write(json);
